Scrollbar migration guide (#5667)
flutter/website: src/docs/release/breaking-changes/default-desktop-scrollbars.md

---
title: Default Scrollbars on Desktop
description: ScrollBehaviors will now automatically build Scrollbars on Desktop platforms.
---

## Summary

> `ScrollBehavior`s will now automatically apply `Scrollbar`s to

Reviewer: Delete "will".

scrolling widgets on desktop platforms - Mac, Windows and Linux.

## Context

Prior to this change, `Scrollbar`s were applied to scrolling widgets
manually by the developer across all platforms. This did not match
developer expectations when executing Flutter applications on desktop platforms.

> Now, the inherited `ScrollBehavior` apply a `Scrollbar` automatically

Reviewer: ```suggestion
Now, the inherited `ScrollBehavior` applies a `Scrollbar` automatically
```

> Now, the inherited `ScrollBehavior` will apply a `Scrollbar` automatically

Reviewer: will apply -> applies

To provide better management and control of this feature, `ScrollBehavior`
has also been updated. The `buildViewportChrome` method, which applied
a `GlowingOverscrollIndicator`, has been deprecated. Instead, `ScrollBehavior`
now supports individual methods for decorating the viewport, `buildScrollbar`
and `buildOverscrollIndicator`. These methods can be overridden to control
what is built around the scrollable.

Further more, `ScrollBehavior` subclasses `MaterialScrollBehavior` and
`CupertinoScrollBehavior` have been made public, allowing developers to extend
and build upon the other existing `ScrollBehavior`s in the framework. These
subclasses were previously private. 

## Description of change

The previous approach called on developers to create their own `Scrollbar`s on
all platforms. In some use cases, a `ScrollController` would need to be provided
to the `Scrollbar` and the scrollable widget.

<!-- skip -->
```dart
final ScrollController controller = ScrollController();
Scrollbar(
  controller: controller,
  child: ListView.builder(
    controller: controller,
    itemBuilder: (BuildContext context, int index) {
      return Text('Item $index');
    }
  )
);
```

> The `ScrollBehavior` will now apply the `Scrollbar` automatically

Reviewer: will now apply -> now applies

> when executing on desktop, and will handle providing the `ScrollController`

Reviewer: will handle -> handles

to the `Scrollbar` for you.

<!-- skip -->
```dart
final ScrollController controller = ScrollController();
ListView.builder(
  controller: controller,
  itemBuilder: (BuildContext context, int index) {
   return Text('Item $index');
 }
);
```

Some widgets in the framework are exempt from this automatic `Scrollbar` application. They
are:
- `EditableText`, when `maxLines` is 1.
- `ListWheelScrollView`
- `PageView`
- `NestedScrollView`

Since these widgets manually override the inherited `ScrollBehavior`
to remove `Scrollbar`s, all of these widgets now have a `scrollBehavior`
parameter so that one can be provided to use instead of the override.

This change did not cause any test failures, crashes, or error messages
in the course of development, but it may result in two `Scrollbar`s
being rendered in your application if you are manually adding `Scrollbar`s
on desktop platforms.

If you are seeing this in your application, there are several ways to
control and configure this feature.

> - Remove manually applied `Scrollbar`s in your application when running on desktop.

Reviewer: ```suggestion
- Remove the manually applied `Scrollbar`s in your application when running on desktop.
```

- Extend `ScrollBehavior`, `MaterialScrollBehavior`, or `CupertinoScrollBehavior`
to modify the default behavior.
  
    - With your own `ScrollBehavior`, you can apply it app-wide by setting

>       `MaterialApp.scrollBehavior`, or `CupertinoApp.scrollBehavior`.

Reviewer: ```suggestion
      `MaterialApp.scrollBehavior` or `CupertinoApp.scrollBehavior`.
```

    - Or, if you wish to only apply it to specific widgets, add a
      `ScrollConfiguration` above the widget in question with your

>       custom `ScrollBehavior`

Reviewer: ```suggestion
      custom `ScrollBehavior`.
```

> Your scrollable widgets then inherit this and reflect this behavior instead.

Reviewer: ```suggestion
Your scrollable widgets then inherits this and reflects this behavior.
```

> Your scrollable widgets will then inherit this and reflect this behavior instead.

Reviewer: will then -> then
